test/e2e/testsuites/dynamically_provisioned_pod_toleration.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testsuites
+
+import (
+	"context"
+	"strings"
+
+	"github.com/onsi/ginkgo"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/controller"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+)
+
+//  will provision required PV(s), PVC(s) and Pod(s)
+// Pod should successfully be re-scheduled on failover in a cluster with AzDriverNode and AzVolumeAttachment resources
+type PodToleration struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    PodDetails
+	Volume                 VolumeDetails
+	AzDiskClient           *azDiskClientSet.Clientset
+	StorageClassParameters map[string]string
+}
+
+func (t *PodToleration) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	_, maxMountReplicaCount := azureutils.GetMaxSharesAndMaxMountReplicaCount(t.StorageClassParameters)
+
+	// Get the list of available nodes for scheduling the pod
+	nodes := ListNodeNames(client)
+	if len(nodes) < maxMountReplicaCount+1 {
+		ginkgo.Skip("need at least %d nodes to verify the test case. Current node count is %d", maxMountReplicaCount+1, len(nodes))
+	}
+
+	ctx := context.Background()
+
+	// set node taint
+	numNodesWithTaint := len(nodes) - (maxMountReplicaCount + 1)
+	nodesWithTaint := map[string]struct{}{}
+	for i := 0; i < numNodesWithTaint; i++ {
+		nodeObj, err := client.CoreV1().Nodes().Get(ctx, nodes[i], metav1.GetOptions{})
```

Comment:
We're labelling or tainting nodes in the order returned by `ListNodeNames` which includes the master node. Given the master node is already tainted to prevent schedule workloads (by default), `ListNodeNames` should filter out nodes with the `node-role.kubernetes.io/master` label (or only include nodes with the `node-role.kubernetes.io/agent` label), and probably be renamed to `ListAgentNodes`.